File: sdk/go/ggid/middleware.go
package ggid

import (
	"context"
	"net/http"
	"strings"
)

// Middleware returns an http.Handler that authenticates requests via GGID JWT.
// Public paths (login, register, healthz) skip authentication.
func (c *Client) Middleware(next http.Handler) http.Handler {
	var publicPaths = map[string]bool{
		"/": true, "/healthz": true, "/docs": true,
		"/api-docs": true, "/login": true, "/register": true,
	}

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		path := r.URL.Path

		// Skip public paths and auth endpoints
		if publicPaths[path] || strings.HasPrefix(path, "/api/v1/auth/") ||
			strings.HasPrefix(path, "/oauth/") {
			next.ServeHTTP(w, r)
			return
		}

		// Extract Bearer token
		authHeader := r.Header.Get("Authorization")
		if !strings.HasPrefix(authHeader, "Bearer ") {
			http.Error(w, `{"error":"missing bearer token"}`, http.StatusUnauthorized)
			return
		}

		token := authHeader[7:]

		// Verify token
		claims, err := c.VerifyToken(r.Context(), token)
		if err != nil {
			status := http.StatusUnauthorized
			if err == ErrTokenExpired {
				http.Error(w, `{"error":"token expired"}`, status)
			} else {
				http.Error(w, `{"error":"invalid token"}`, status)
			}
			return
		}

		// Inject claims into request context
		ctx := context.WithValue(r.Context(), claimsKey{}, claims)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// claimsKey is the context key for JWT claims.
type claimsKey struct{}

// ClaimsFromContext extracts JWT claims from the request context.
func ClaimsFromContext(ctx context.Context) map[string]interface{} {
	if v, ok := ctx.Value(claimsKey{}).(map[string]interface{}); ok {
		return v
	}
	return nil
}

// RequirePermission returns a middleware that checks if the user has permission
// for the given resource and action.
func (c *Client) RequirePermission(resource, action string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			claims := ClaimsFromContext(r.Context())
			if claims == nil {
				http.Error(w, `{"error":"not authenticated"}`, http.StatusUnauthorized)
				return
			}

			// Get token from header
			authHeader := r.Header.Get("Authorization")
			token := ""
			if strings.HasPrefix(authHeader, "Bearer ") {
				token = authHeader[7:]
			}

			result, err := c.CheckPermission(r.Context(), token, resource, action)
			if err != nil || !result.Allowed {
				http.Error(w, `{"error":"forbidden"}`, http.StatusForbidden)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

File: sdk/ruby/lib/ggid/middleware.rb
# frozen_string_literal: true

module GGID
  # Middleware module — Rails ActionController helpers.
  #
  # In your ApplicationController:
  #
  #   class ApplicationController < ActionController::Base
  #     include GGID::Middleware
  #
  #     before_action :require_auth
  #
  #     # For specific controllers:
  #     # before_action -> { require_permission!('products', 'read') }, only: [:index]
  #     # before_action -> { require_role!('admin') }, only: [:admin_panel]
  #
  #     def current_ggid
  #       @ggid ||= GGID::Client.new(
  #         base_url: Rails.configuration.x.ggid_base_url,
  #         tenant_id: Rails.configuration.x.ggid_tenant_id,
  #       )
  #     end
  #   end
  module Middleware
    # Extract the bearer token from the Authorization header.
    #
    # @return [String, nil]
    def ggid_token
      auth_header = request.headers["Authorization"].to_s
      match = auth_header.match(/^Bearer\s+(.+)/i)
      match ? match[1].strip : nil
    end

    # Get the verified claims from the current request's token.
    #
    # @return [Claims, nil]
    def ggid_claims
      return @ggid_claims if defined?(@ggid_claims)
      token = ggid_token
      return @ggid_claims = nil unless token
      begin
        @ggid_claims = current_ggid.verify_token(token)
      rescue InvalidTokenError
        @ggid_claims = nil
      end
    end

    # Require authentication — redirects or returns 401 if no valid token.
    # Use as: before_action :require_auth
    def require_auth
      token = ggid_token
      unless token
        render_unauthorized("Missing or invalid Authorization header")
        return
      end
      begin
        @ggid_claims = current_ggid.verify_token(token)
        @ggid_access_token = token
      rescue InvalidTokenError => e
        render_unauthorized(e.message)
      end
    end

    # Require a specific permission.
    # Use as: before_action -> { require_permission!('products', 'read') }
    #
    # @param resource [String]
    # @param action [String]
    def require_permission!(resource, action)
      token = ggid_token
      unless token
        render_unauthorized("Authentication required")
        return
      end
      result = current_ggid.check_permission(token, resource, action)
      unless result.allowed
        render_forbidden("Permission denied: #{resource}:#{action}")
      end
    end

    # Require a specific role.
    # Use as: before_action -> { require_role!('admin') }
    #
    # @param role [String]
    def require_role!(role)
      token = ggid_token
      unless token
        render_unauthorized("Authentication required")
        return
      end
      begin
        claims = current_ggid.verify_token(token)
      rescue InvalidTokenError => e
        render_unauthorized(e.message)
        return
      end
      unless claims.has_role?(role)
        render_forbidden("Required role: #{role}")
      end
    end

    # Convenience: check if the current user has a permission (non-raising).
    #
    # @param resource [String]
    # @param action [String]
    # @return [Boolean]
    def can?(resource, action)
      token = ggid_token
      return false unless token
      result = current_ggid.check_permission(token, resource, action)
      result.allowed
    rescue StandardError
      false
    end

    # Convenience: check if the current user has a role (non-raising).
    #
    # @param role [String]
    # @return [Boolean]
    def has_role?(role)
      claims = ggid_claims
      claims&.has_role?(role)
    end

    private

    def render_unauthorized(message)
      respond_to do |format|
        format.json { render json: { error: "unauthorized", message: message }, status: :unauthorized }
        format.html { redirect_to "/login", alert: message }
        format.any { render plain: message, status: :unauthorized }
      end
    end

    def render_forbidden(message)
      respond_to do |format|
        format.json { render json: { error: "forbidden", message: message }, status: :forbidden }
        format.html { render plain: "403 Forbidden: #{message}", status: :forbidden }
        format.any { render plain: message, status: :forbidden }
      end
    end
  end
end
